### ISSUE TITLE
Extract Admin UI from the main repo

### ISSUE THREAD

**Issue by gabor-boros** [rethinkdb/rethinkdb#6864]
It would be great if we would extract the admin ui from this repository. The build would have less dependencies and the admin UI could be release independently.

**Comment by mishawakerman** [comment]
I might take a look at this as it's relevant for our team. Would also be good to talk to the folks that started building the RebirthDB admin console for any tips.

**Comment by gabor-boros** [comment]
FYI: @sagivf and @oferitz

**Comment by oferitz** [comment]
Hi, I'm the one that started this together with @sagivf 2 years ago and unfortunately had to abandon this project due to lack of time. 
2 years is a long time in terms of a Front End project development and i believe that whoever wants continue with this project should consider writing it from scratch, as a lot of the concepts and tech here is a bit outdated and could be done much better (i.e not sure Electron is still the right choice). 
I'm willing to help and even contribute from time to time, just don't have the resources to lead this. Please keep me in the loop for any new updates regrading this.

**Comment by sagivf** [comment]
Hey guys.
I to am pretty busy with a new Job (and family matters), so I can't really lead this either.
I'd be happy to help here and there, so keep me in the loop as well.
(Of course 

As far a the status is concerned - we got quite a lot done (far from what is needed, but still), so I'm not sure, but I probably wouldn't completely abandon it.

@oferitz what would you use instead of electron? 
From what I recall,  setting it up and working with it was a bit of a pain, it stooped working after updating things, or reinstalling on different machines.

So It would be great for it to be a simple web app (create-react-app), however i don't see how you would easily run the underlying queries without something like electron. 

In any case I would recommend starting fresh and copying in parts from what we've done.

**Comment by gabor-boros** [comment]
Since nobody volunteered for leading this, I can take it.

I'd say we could start a new non-electron project with Typescript (using rethinkdb-ts), nuxt and tailwindcss/tailwindui (for the sake of simplicity) from scratch. Although I'm not a typescript or even front-end guru, I can assign some time for that. Why not electron? Maintaining one binary release for X platforms is enough for now I think.

What do you think about joining to this initiative @sagivf, @oferitz and @mishawakerman. Maybe this could be interesting for @adriantodt, @ronzeidman and even @atassis. I know that everyone is busy so no commitments at all 😇

**Comment by atassis** [comment]
@gabor-boros I can join as well because right now we use only query execution UI, all other views are more useful in our datadog integration. I do the frontend part of our organization as well and can help with this. I try not to spend much time for hobby like things, but it seems it is still more than enough per person for rethinkdb.

**Comment by atassis** [comment]
And yep, I have worked already with electron and other stuff, and have experience in different environments in JS, my experience might be useful for you guys.

**Comment by adriantodt** [comment]
@gabor-boros My HTML/CSS experience goes as far as W3Schools, using Bootstrap 2/3/4, using MaterializeCSS and using Bulma. Never tinkered with Electron, but I know how to JS quite a bit (assuming I can get my JS IDE to work with whatever target)

**Comment by gabor-boros** [comment]
> And yep, I have worked already with electron and other stuff, and have experience in different environments in JS, my experience might be useful for you guys.


Oh, actually I meant not doing the admin UI in electron 😅

**Comment by atassis** [comment]
Why not?

**Comment by gabor-boros** [comment]
Discussing on the issue could be beneficial for transparency 😇

My only reasoning is that we could serve the new admin through the rethinkdb binary as we are doing now (but pulled as an external dependency).

Last time we decided to have electron based app, so I can be convinced again ☺️

**Comment by atassis** [comment]
To keep rethinkdb db SOLID and install the admin if needed. It is an explicit dependency for the production, which can be connected easily if needed. We can still do not an electron app but the external admin UI. No modern and actual database serves integrated UI inside.

**Comment by gabor-boros** [comment]
Ok, that would reduce the build complexity of the DB itself what I think everyone would appreciate, especially @srh (sorry for looping you in it’s a FYI for you that we are planning to remove the admin UI too).

The question is the tech stack now. As I mentioned earlier I’m not a front end developer so my knowledge about that is limited to the latest mainstream options on the market.

I would choose Vuejs + Tailwind but I can be convinced about React too. 

I know it’s against my previous comment, but if we would do a react app it would be easier for the end users (maybe).

@atassis Have you worked with electron + react/vue previously in production?

**Comment by gabor-boros** [comment]
Last time the painful setup and maintenance of electron had negative effects tho. Maybe it can be handled better or we could choose an other option (if there is any)

**Comment by atassis** [comment]
I can handle it, it is not bad as you might think. We just need to prepare the specification and I'll try my best.

**Comment by atassis** [comment]
Also I know React well, as well as the complete react stack. Truly I can do any JS related stuff and know Devops a little bit

**Comment by gabor-boros** [comment]
@atassis The specification would be to have the same functionality (excluding logs which cannot be transferred) as we have now. Thankfully, the current admin ui is a great base for the new one in terms of we don’t have to figure out the queries used there.

What would you use as a front end framework with react? I’ll create an empty repo for this by the way.

**Comment by atassis** [comment]
That might be nice. Still, it is better to setup a more precise specification, but I think I got all the needed information to make it by myself, so you can agree or not with that. Then I can try doing something. For example, we can setup an admin which can be build either in electron or in the docker image. Has to be researched a little bit since I have written my last code in electron several years ago.

**Comment by mwakerman** [comment]
I was against this being an electron app last time and still am for a couple of reasons:

- Seperate applications (electron or otherwise) can be difficult to get installed in an enterprise environment. I love that the best RethinkDB query GUI is bundled with the database.
- As far as I know, Electron apps can’t be used iOS, Android, or Chrome OS. This isn’t a deal breaker but is something about the current UI that’s saved me a couple of times (being able to debug an issue while away from a computer).
- Electron is a fairly large framework and from my limited experience using it (but many blog posts attest to this) you want people who know it well to be involved to prevent performance and other issues. Ultimately it’s a significant dependency that will require expertise and this project doesn’t have anyone committed to it full time, let alone someone with Electron experience.

I don’t have a line of sight to what the admin UI needs but i do wonder if this can be a purely frontend application (static site) that you point at a RethinkDB instance, login/upload cert, and off you go. For us, only being able to run it as the admin user  has/is causing us issues with audit and compliance obligations.

**Comment by atassis** [comment]
@mwakerman I have also mentioned creating the docker image for that. I suppose it can't be a SPA since it uses node API. I'll repeat, I want to setup an app which can be distributed both as an electron and a docker image.

**Comment by mwakerman** [comment]
That sounds reasonable. Which part of the Node API would the app need? From memory, rethinkdbdash had browser support but I believe it was dropped for rethinkdb-ts with a comment that it could be restored.

**Comment by adriantodt** [comment]
How about doing it the pgAdmin 4 way? A standalone webserver which can be turned on and off independently of the RethinkDB itself. This way it wouldn't lose the web UI functionality and still making it independent from the server itself.

**Comment by sagivf** [comment]
- Although I to likes the UI being bundled in, I think decoupling the project it important for maintainability.

- Regarding electron. I may be missing something, but I think the main motivation was having it be able to run the queries locally like this - https://github.com/rethinkdb/rethinkdb-desktop/blob/master/app/main/db/queries/database.js.
If  I understand correctly without electron you would need to add a REST or socket layer to serialize and transfer queries and data. If that's not an issue it would indeed be much simpler to write this as standalone web app and bundle it into a native app later if needed.

- As far as frameworks go React would be my go to, but I would let who ever is contributing the most make that decision.

**Comment by gabor-boros** [comment]
@atassis How would the docker and electron based version look like? I only made some hobby stuff in electron so I’m not that familiar with it (though I can pick it up quickly if needed).

@sagivf For me, the most important is to use a tech stack (including react or vue doesn’t really matter for me), which can be maintained by the majority of the community. If that’s react I’m fine with that too.

The only thing I’d stick to is to use Typescript.

The idea of decoupling the admin ui would be pretty useful, since that would remove lot of dependencies from the DB itself.

Tbh, it is extremely good sometimes that the Admin UI is built in the DB, but causing some issues like you cannot release a ui update easily without releasing the DB too which is like a day of work to prepare (if all the tests are passing and so on).

Also, we could add features like connecting to multiple databases from the same window in the future (even electron or SPA). For now it would just blow up the scope I think.

**Comment by atassis** [comment]
@gabor-boros Just the simple node js server built in the electron. Electron will be just needed to proxy client requests inside the server. I'll do that as well.

**Comment by gabor-boros** [comment]
@atassis As far as I understand it will be enough to have one repo for this, but correct me if I’m wrong. I’ll create the repo soon.

**Comment by gabor-boros** [comment]
FYI: @atassis Created the repo called rethinkdb-admin.

What do you think @sagivf, should we archive the rethinkdb-desktop repo? I would say yes based on your comment about the status of that.

**Comment by atassis** [comment]
@gabor-boros yep, it is better to keep as few repositories for the code as possible. They must be strict but laconic (SOLID). I'll clone and try to do the WIP in next week or two.

**Comment by NetOperatorWibby** [comment]
I have strong negative views on React and would prefer this new admin UI to be built with Svelte. Of course, this is just one person's preference and I will most likely build this myself at some point anyway.

Like others have mentioned above I lack time but not due to being busy. I've just found it difficult to work on my own projects due to this pandemic.

**Comment by atassis** [comment]
@NetOperatorWibby I strongly discourage using svelte, this is an instrument which is not used widely. React is the best up here also because of its strong community base.

**Comment by NetOperatorWibby** [comment]
Being an incumbent is perhaps React's best feature and there are undoubtedly tons of people who are familiar with it. We have a chance of being on the cutting edge with Svelte but I understand transitioning to something new is terrifying for some.

**Comment by sagivf** [comment]
> FYI: @atassis Created the repo called rethinkdb-admin.
> 
> What do you think @sagivf, should we archive the rethinkdb-desktop repo? I would say yes based on your comment about the status of that.

Archive mean the code is still available right? If so, sure, go ahead.

I have to say it still isn't clear to me how the actual querying of the database is going to be done without electron. I assume some sort of node server and a REST API, but maybe i'm missing something.

**Comment by atassis** [comment]
@sagivf Yep, I am going to create a single mechanism for querying from frontend for both electron and node API. It will do different work depending on for which env it was built. Something alike slack, skype, etc nowadays

**Comment by gabor-boros** [comment]
@NetOperatorWibby Before opening this discussion (or so), I've checked Svelte too. I'm not against the idea, and I'd not like to close any possibilities. It is a lot faster than React or Vue (almost faster than any JS frameworks nowadays except two or three). My biggest concern about Svelte is the small community around that compared to its competitors. This wouldn't mean anything bad, but I'm afraid of building community driven project (Admin UI) on a framework which is known by (hash guessing) 1/100 in the RethinkDB community. 

If I would be a sole developer or we would have a 5 person team commited to work on this (not full time, but having time and continuous interest in building this) I'd go with Svelte too if it can support types. For the latter my reasoning is that - during my not to deep adventure in the front end world - most of the issues are coming from lack of types and poor code quality.

@sagivf Yes, this is exactly what I meant. I'll archive the repo then.

@atassis It sounds good to me to have something like this and not need to maintain 2 different things.

**Comment by NetOperatorWibby** [comment]
I attended a talk today where Rich Harris demo’d Svelte and he confirmed that types are coming to Svelte. That’s not gonna happen anytime soon though.

As a (relatively) recent convert to TypeScript, I absolutely love ways to prevent myself from making mistakes and would feel safer with an admin interface to the lifeline of my project being strongly typed.

**Comment by ChrisTalman** [comment]
I would also recommend Svelte. It certainly isn't used as widely as other frameworks. However, the ways in which it is different from other frameworks helps a lot to offset its lack of popularity. I don't favour it because of its performance, but rather its syntax. Unlike other frameworks, it largely sticks with what Web developers already know: HTML and JavaScript. In this sense, it's a lot more intuitive, and thus should be comparatively easier to understand and learn for newcomers. By contrast, other frameworks are a lot less intuitive, and require developers to learn a lot more about idiosyncratic syntax and concepts. Thus, while Svelte doesn't have the popularity of other frameworks, in critical regards it doesn't actually need it.

**Comment by ChrisTalman** [comment]
An alternative to Electron just came to mind. I'm not sure whether this has been discussed before.

The UI could be entirely extracted from the database binary, thus allowing it to be iterated and released separately. However, instead of it running in Electron, it could be deployed at https://console.rethinkdb.com as a single page app (SPA). Moreover, it could be made progressive Web app (PWA) compliant. This would enable it to be installed on the homescreens of desktops, mobiles, and tablets. This would also allow it to be used offline, which could be useful within an intranet which is separated from the Internet. Users could also host it themselves, if they could not or preferred not to use https://console.rethinkdb.com.

The primary issue with a standard Web app is that it cannot communicate directly with the RethinkDB driver port. This is because there is currently no Web API to open a raw TCP connection. That is presumably one of the primary reasons why Electron was favoured in the past, as it provides an API to do this. However, one way to work around this issue could be to maintain a minimal proxy server in the database binary. Its sole purpose would be to translate from HTTPS to the database driver socket. While this would mean that the database binary would not be entirely divorced from the UI as a whole, it would seem to mean that there is far less to maintain within the binary, and the UI itself is able to be iterated and released independently.

**Comment by atassis** [comment]
@ChrisTalman there is no reason to use an instrument with so small community as svelte has. There best library for UI here is react, less attractive but still attractive choice is angular. Others are having too small communities, we can't allow that for rethinkdb, no matter how good they are. The faster contributors start developing something- the better.
And, by the way, I have started to create a minimal approach for an admin, will try to finish alpha version in this month

**Comment by ChrisTalman** [comment]
@atassis I provided a reason in my first comment.

To elaborate upon that reason, Svelte is much more intuitive than other frameworks. This is because it essentially just uses HTML and JavaScript. Further, it was designed to be minimal in scope compared with other frameworks. As a result, it's a lot easier to get on board with. Thus, while fewer people know it, it isn't difficult for them to get into. And that's important for contributions. If you use a framework which is more popular but also more idiosyncratic, like React or Angular, you risk putting off contributors who are not familiar with it. There's a lot of value in Svelte in this regard.

**Comment by mwakerman** [comment]
@ChrisTalman you’re not wrong that Svelte looks easier to learn and contribute with if your starting point is knowledge of HTML and JavaScript.

However in this case, I think React is the right call. It’s community is huge (I’d guess that most Svelte developers also know React), it’s backed by a large company (despite what we may think of them), it’s still improving in performance and usability at a fast pace, and there are a huge number of high-quality, battle tested component libraries that we can leverage. I’m likely going to be a small contributor to this project so weight my opinion accordingly but wanted to have it said.

As a weird aside, I’ve interviewed some grads that know React but don’t “know” JavaScript 😅

**Comment by McSneaky** [comment]
After using both Svelte and Vue in prod in teams I'd stick with Vue, especially for community project that needs maintainers
Svelte has tendency to turn spaghetti real fast, when multiple devs are working on things and app gets more complex. Introducing communication between components, state machines etc
But that's something that can be avoided with good engineering decisions so I wouldn't add any weight to that argument

Bigger problem is community, finding some edge case solution in Svelte is a lot harder than in Vue. Or finding developers who use Svelte is harder than finding devs who have Vue experience (according to stateofjs Svelte has over 10x less devs familiar with it compared to React)

Imagine that RethinkDB community is about 100 people (who are active in Discord, Slack, Github etc), let's say about 50 of them are familiar with frontend, then out of those 50 35 know React, 20 Vue and 3 Svelte (taking stateofjs stats again). How many are actually active community members I have no idea. but I doubt we could find many people who actually have time to contribute, so numbers might be even lower (hopefully higher)

Vue and Svelte both have quite the same learning curve, which is a lot smaller than React or Angular

Not saying that Svelte is bad or anything. It's really good and I like to use it, but wouldn't build real-time database management system on top of it and hoping that I can find some other people to help me out from small community like RethinkDB one is..
I really-really help that community grows tho :yum:

Choosing React won't be a bad choice. It might be slower* than Svelte, but it has a lot more people who are familiar with it, a lot more people who might have time to contribute etc

Even tho React would be last option I'd choose if I'd have to solo build anything. Just because I've had to fix too much too broken React code and it has left really bad taste - don't even want to see React for some time now :sweat: 

_*Won't matter much if framework can update 1m DOM nodes in a sec or 1.3m nodes. Something is really off if you really need faster framework to update millions of DOM nodes..._

**Comment by NetOperatorWibby** [comment]
React is crufty and as time goes on and native JavaScript matures, developers are moving away from it (anecdotally, of course). IMHO, Vue's learning curve is much steeper than Svelte. With Svelte, if you know how to code a webpage you're halfway there.

There's far less setup to get started and the development experience is a breath of fresh air.

This conversation has given me motivation to start working on it myself. Judging by the way the convo is going, have fun with React.

**Comment by ChrisTalman** [comment]
> Svelte has tendency to turn spaghetti real fast
@McSneaky Not to get too off topic, but, out of curiosity, have you found there to be any difference between Svelte 2 and 3 in this regard?

In 2, it was possible to have all your behaviour outside the component file. In 3, a lot less can be done outside the component file. This was a shame for me personally, as I found that the code structure could be much better organised and easier to reason about in 2 because of this.

**Comment by McSneaky** [comment]
Sorry, I've only been working with v3, haven't used previous versions, don't know about those :(

**Comment by NetOperatorWibby** [comment]
Oh, another option I thought of is [Stimulus.js](https://stimulusjs.org). It's from the people behind Basecamp. It's a little bit more of a learning curve than Svelte but I enjoy it as well.

**Comment by mtomov** [comment]
> Being an incumbent is perhaps React's best feature

haha, that made my day. Well said.

Many experienced devs are plain tired and bored of the jumping through the hoops that React demands, and indeed looking at Svelte - they've thought about many of those shortcomings. If you look at some of the articles:

- https://svelte.dev/blog/svelte-3-rethinking-reactivity
- https://svelte.dev/blog/write-less-code#State (the whole concept of hooks overhauled)

I think such large projects can serve as examples to the community, so newcomers can see that there's more to frontend code than just React. Especially with open source projects, learning is the bigger motivator for people. 

Stimulus is great, but you'd likely need a server to send out html for it.

**Comment by cope** [comment]
I've developed and am still maintaining an app for work with Electron and for the most part I really liked it. But, maybe something more lightweight like Neutralinojs could be an alternative? The [comparison](https://github.com/neutralinojs/evaluation) seems promising, excluding the "_Windows edition is based on IE_" at the end.

I would love to contribute to rethinkdb-admin, regardless of React vs Svelte discussion.

Question, since the [rethinkdb-admin](https://github.com/rethinkdb/rethinkdb-admin) repo exists, maybe these decisions/discussions on what to use for it should be moved there? :)

**Comment by bajro17** [comment]
Is flutter one of the options it work on all devices?

**Comment by lsabi** [comment]
@atassis I think you are quite wrong saying that svelte has a small community. I would judge quality over quantity and I can say that they are quite active. Yes, svelte has "only" 36.5k stars, while vue and react have, respectively, 171k and 155k, but I remember that not so long ago vue had 30k as well. Also, react has 1496 contributors, vue has 371 contributors and svelte has 323 contributors. That means that there is one contributor of react, vue and svelte for every 114, 417 and 113 stars, respectively.

Numbers apart, I believe it's more a gamble on the future and not just on the present. Some years ago there was angular, then react, then vue and then ... which one? It's true that choosing a framework that's too young is likely to lead to problems, but also choosing a framework that's too old will lead to problems (I mean, who uses perl anymore?).

That being said, I believe it's optimal to choose a framework that is both mature and easy to learn. The former is self explanatory, while the latter aims at involving new users that do not know that framework yet. I'm a back end guy, but I tried to learn react, angular and vue, giving up because they were way too complex and the tutorials way too long in order to do anything that wasn't 1+1 = 2. It took me 5 minutes to get started with svelte and half a day to go through all of it. There has to be considered the possibility that an angular dev (or another framework) may be willing to try something new, possibly learning from this project and, I hope, contribute to it once they have learned enough. Instead, if there is a really complex framework, it'll be harder to find new devs once the hype for that framework will be over. I haven't used svelte for really complex things, so I can't say much, but for what I've used it (I learned nodejs with svelte) it wasn't difficult to build an MVP.

Maybe @mtomov can back me up on this as I don't have gitter, so I'm not in the community and I'm not a front end guy.

All in all, I would write on a virtual board all pros and cons that are emerging in this issue in order to see the big picture and include how many devs willing to help are front devs and how many know which framework(s).

Also, since I'vent thrown enough gasoline on the fire, why not consider [nwjs.io](https://nwjs.io/) as alternative to electron? @cope do you have any experience with it?

**Comment by cope** [comment]
@lsabi no, I only worked with Electron so far. But, if I try anything next, it will certainly be [neutralino](https://neutralino.js.org/)...

**Comment by srh** [comment]
Regarding build dependencies, the admin UI is already made from pre-generated web assets.  So that's more a matter of fixing the build system so it doesn't spuriously build nodejs.

Regarding what platform to move the admin UI to, or how to build an alternative one, it doesn't matter unless you actually build your own desktop client.  Whoever builds it gets to build it however they want.

**Comment by srh** [comment]
It's also worth noting that if every way of making a desktop app will become obsolete after 2 years, you're doomed to be using so-called outdated tech.  So maintaining the existing app is probably the least amount of work.

**Comment by atassis** [comment]
Guys, there is already a repo with an admin making minimal required things that current admin page does, but I am not having enough time making it solo. Would like someone else to join the development to help me understand how to create a better UX admin panel. Do not even need help in writing the code, need a help with structuring pages in opinionating what to do and in which way.
